### PR TITLE
Correct warnings that contains Unicode message

### LINKED ISSUE #2436: UnicodeEncodeError in warnings capture
- [x] Include a detailed description of the bug or suggestion
We run tests in a CI server, but somethings it failed because the below internal error:

```text
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/main.py", line 105, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/main.py", line 141, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
INTERNALERROR>     _MultiCall(methods, kwargs, hook.spec_opts).execute()
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 613, in execute
INTERNALERROR>     return _wrapped_call(hook_impl.function(*args), self.execute)
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 250, in _wrapped_call
INTERNALERROR>     wrap_controller.send(call_outcome)
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 280, in get_result
INTERNALERROR>     _reraise(*ex)  # noqa
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 265, in __init__
INTERNALERROR>     self.result = func()
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/main.py", line 164, in pytest_runtestloop
INTERNALERROR>     item.config.hook.pytest_runtest_protocol(item=item, nextitem=nextitem)
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
INTERNALERROR>     _MultiCall(methods, kwargs, hook.spec_opts).execute()
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 613, in execute
INTERNALERROR>     return _wrapped_call(hook_impl.function(*args), self.execute)
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/vendored_packages/pluggy.py", line 250, in _wrapped_call
INTERNALERROR>     wrap_controller.send(call_outcome)
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/warnings.py", line 73, in pytest_runtest_protocol
INTERNALERROR>     yield
INTERNALERROR>   File "/usr/local/lib/python2.7/contextlib.py", line 24, in __exit__
INTERNALERROR>     self.gen.next()
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/_pytest/warnings.py", line 66, in catch_warnings_for_item
INTERNALERROR>     warning.filename, warning.lineno, warning.line)
INTERNALERROR>   File "/buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/warnings.py", line 50, in formatwarning
INTERNALERROR>     s =  "%s: %s: %s\n" % (lineno, category.__name__, message)
INTERNALERROR> UnicodeEncodeError: 'ascii' codec can't encode characters in position 35-37: ordinal not in range(128)
```

This problem is weird, sometimes happened and sometimes not. So I don't know why, and I can't let it happened again manually.

- [x] `pip list` of the virtual environment you are using

```
amqp (2.1.4)
appdirs (1.4.3)
backports-abc (0.5)
better-exceptions (0.1.8)
billiard (3.5.0.2)
celery (4.0.2)
certifi (2017.4.17)
common (2.4.8)
contextlib2 (0.5.5)
coverage (4.4.1)
decorator (4.0.11)
enum34 (1.1.6)
funcsigs (1.0.2)
gredis (0.0.8)
hashids (1.2.0)
Jinja2 (2.9.6)
kombu (4.0.2)
MarkupSafe (1.0)
mock (2.0.0)
modcov (0.0.3.dev30)
MySQL-python (1.2.5)
olefile (0.44)
packaging (16.8)
pbr (3.0.1)
Pillow (4.1.1)
pip (9.0.1)
py (1.4.33)
py-dateutil (2.2)
pycrypto (2.6.1)
pycurl (7.43.0)
pyparsing (2.2.0)
pytest (3.1.0)
pytest-cov (2.5.1)
pytz (2017.2)
qrcode (5.3)
raven (6.0.0)
redis (2.10.5)
requests (2.14.2)
setuptools (35.0.2)
singledispatch (3.4.0.3)
six (1.10.0)
SQLAlchemy (1.1.10)
sqlalchemy-migrate (0.11.0)
sqlparse (0.2.3)
suds-jurko (0.6)
Tempita (0.5.2)
tornado (4.5.1)
vine (1.1.3)
wheel (0.29.0)
wheezy.captcha (0.1.44)
xlrd (1.0.0)
xlwt (1.2.0)
```

- [x] pytest and operating system versions

```shell
$ .tox/py27/bin/pytest --version
This is pytest version 3.1.0, imported from /buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/pytest.pyc
setuptools registered plugins:
  pytest-cov-2.5.1 at /buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/pytest_cov/plugin.py
  celery-4.0.2 at /buildbot/worker/agent-admin_auto-test/build/.tox/py27/lib/python2.7/site-packages/celery/contrib/pytest.py

$ uname -a
Linux d4ad8bbd6105 4.4.0-78-generic #99-Ubuntu SMP Thu Apr 27 15:29:09 UTC 2017 x86_64 GNU/Linux
```

- [ ] Minimal example if possible

### PR DESCRIPTION
**Pull request by coldnight**
This PR try to fix #2436.

- [x] Target: `master`;

Unless your change is trivial documentation fix (e.g.,  a typo or reword of a small section) please:

- [x] Make sure to include one or more tests for your change;
- [ ] Add yourself to `AUTHORS`;
- [x] Add a new entry to `CHANGELOG.rst`
  * Choose any open position to avoid merge conflicts with other PRs.
  * Add a link to the issue you are fixing (if any) using RST syntax.
  * The pytest team likes to have people to acknowledged in the `CHANGELOG`, so please add a thank note to yourself ("Thanks @user for the PR") and a link to your GitHub profile. It may sound weird thanking yourself, but otherwise a maintainer would have to do it manually before or after merging instead of just using GitHub's merge button. This makes it easier on the maintainers to merge PRs.

### CODE REVIEW

File: _pytest/warnings.py [hunk 1]
```diff
@@ -62,7 +64,7 @@ def catch_warnings_for_item(item):
 
         for warning in log:
             msg = warnings.formatwarning(
-                warning.message, warning.category,
+                compat.safe_str(warning.message), warning.category,
```
Comment (RonnyPfannschmidt) [inline _pytest/warnings.py]: its required to warn the user here that his warning message is broken as it is not a str instance (after all this is a stdlib problem wrokaround)

Comment (nicoddemus) [inline _pytest/warnings.py]: Sounds good.

File: testing/test_warnings.py [hunk 2]
```diff
@@ -106,3 +109,29 @@ def test_ignore(testdir, pyfile_with_warnings, method):
     ])
     assert WARNINGS_SUMMARY_HEADER not in result.stdout.str()
 
+
+
+def test_unicode(testdir, pyfile_with_warnings):
+    testdir.makepyfile('''
```
Comment (nicoddemus) [inline testing/test_warnings.py]: Excellent work! Please add a CHANGELOG entry as well. 👍

Comment (coldnight) [inline testing/test_warnings.py]: Sorry, I don't know which section should I add the entry to CHANGELOG, could you please tell me that?

Comment (nicoddemus) [inline testing/test_warnings.py]: Sorry did not realize we did not prepare the CHANGELOG for new entries.

Just something like this to the CHANGELOG:

```
3.1.1 (unreleased)
==================

* Fix encoding errors for unicode warnings in Python 2.
```

We are introducing towncrier to manage the CHANGELOG in #2431, but we will take care of moving the entry to that format later! 👍

**Comment by coldnight** [comment]
@RonnyPfannschmidt @nicoddemus r?

**Comment by nicoddemus** [comment]
Excellent work, thanks a ton! 👍 

Btw, would you like to be added to the AUTHORS file? And if so by which name?

**Comment by coldnight** [comment]
@nicoddemus Sure, that's my honor.  Just use `Hui Wang (coldnight)` and thanks for your awesome work on pytest!

**Comment by nicoddemus** [comment]
@coldnight done, thanks again! 😁

**Pull request merged**